src/zabapgit_2fa.prog.abap

```diff
@@ -0,0 +1,659 @@
+      EXPORTING
+        pattern     = iv_supported_url_regex
+        ignore_case = abap_true.
+  ENDMETHOD.
+
+  METHOD authenticate.
+    RAISE EXCEPTION TYPE lcx_2fa_auth_failed. " Needs to be overwritten in subclasses
+  ENDMETHOD.
+
+  METHOD supports_url.
+    rv_supported = mo_url_regex->create_matcher( text = iv_url )->match( ).
+  ENDMETHOD.
+
+  METHOD get_service_id_from_url.
+    rv_id = 'UNKNOWN SERVICE'. " Please overwrite in subclasses
+  ENDMETHOD.
+
+  METHOD is_2fa_required.
+    rv_required = abap_false.
+  ENDMETHOD.
+
+  METHOD delete_access_tokens.
+    RAISE EXCEPTION TYPE lcx_2fa_token_del_failed. " Needs to be overwritten in subclasses
+  ENDMETHOD.
+
+  METHOD raise_internal_error_from_sy.
+    DATA: lv_error_msg TYPE string.
+
+    MESSAGE ID sy-msgid TYPE sy-msgty NUMBER sy-msgno
+            WITH sy-msgv1 sy-msgv2 sy-msgv3 sy-msgv4
+            INTO lv_error_msg.
+    RAISE EXCEPTION TYPE lcx_2fa_auth_failed
+      EXPORTING
+        iv_error_text = |Internal error: { lv_error_msg }| ##NO_TEXT.
+  ENDMETHOD.
+ENDCLASS.
+
+CLASS lcl_2fa_github_authenticator DEFINITION
+  INHERITING FROM lcl_2fa_authenticator_base
+  FINAL
+  CREATE PUBLIC.
+
+  PUBLIC SECTION.
+    METHODS:
+      constructor,
+      get_service_id_from_url REDEFINITION,
+      authenticate REDEFINITION,
+      is_2fa_required REDEFINITION,
+      delete_access_tokens REDEFINITION.
+  PROTECTED SECTION.
+  PRIVATE SECTION.
+    CONSTANTS:
+      gc_github_api_url              TYPE string VALUE `https://api.github.com/`,
+      gc_otp_header_name             TYPE string VALUE `X-Github-OTP`,
+      gc_restendpoint_authorizations TYPE string VALUE `/authorizations`.
+    CLASS-METHODS:
+      set_new_token_request IMPORTING ii_request   TYPE REF TO if_http_request,
+      get_token_from_response IMPORTING ii_response     TYPE REF TO if_http_response
+                              RETURNING VALUE(rv_token) TYPE string,
+      parse_repo_from_url IMPORTING iv_url              TYPE string
+                          RETURNING VALUE(rv_repo_name) TYPE string,
+      set_list_token_request IMPORTING ii_request TYPE REF TO if_http_request,
+      get_tobedel_tokens_from_resp IMPORTING ii_response   TYPE REF TO if_http_response
+                                   RETURNING VALUE(rt_ids) TYPE stringtab,
+      set_del_token_request IMPORTING ii_request  TYPE REF TO if_http_request
+                                      iv_token_id TYPE string,
+      get_authenticated_client IMPORTING iv_username      TYPE string
+                                         iv_password      TYPE string
+                                         iv_2fa_token     TYPE string
+                               RETURNING VALUE(ri_client) TYPE REF TO if_http_client
+                               RAISING   lcx_2fa_auth_failed.
+ENDCLASS.
+
+CLASS lcl_2fa_github_authenticator IMPLEMENTATION.
+  METHOD constructor.
+    super->constructor( 'https?:\/\/(www\.)?github.com.*$' ).
```

Comment:
should be "^https?:\/\/(www\.)?github.com.*$" ? but I guess it does not matter much